feat(Tabs): introduce `orientation` prop

iatanas0v opened the pull request:
[TACO-1038]

### Description

This introduces the `orientation` prop on `Tabs`. This is already supported by MUI itself so I just had to expose it and style the varient.

I bounced this off BASE already - https://toptal-core.slack.com/archives/CCC3GP6CC/p1662703900679159

### How to test

- Check the new vertical tabs example in storybook

### Screenshots

![image](https://user-images.githubusercontent.com/4757057/189835003-396c3ca7-176f-4e1e-8728-9fffb894aaa6.png)


### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Make sure that additions and changes on design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests

**Breaking change**

- n/a ~~codemod is created and showcased in the changeset~~
- n/a ~~test alpha package of Picasso in StaffPortal~~

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[TACO-1038]: https://toptal-core.atlassian.net/browse/TACO-1038?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ozgurkececioglu commented on packages/picasso/src/Tab/Tab.tsx:
```diff
@@ -59,6 +61,7 @@ export const Tab = forwardRef<HTMLDivElement, Props>(function Tab(props, ref) {
       {titleCase ? toTitleCase(label) : label}
     </Typography>
   )
+  const orientation = useContext(TabsOrientationContext) || 'horizontal'
```
Do you still need the default value even after defining it in context creation?

ozgurkececioglu commented:
I think there are some visual differences between implementation and designs we need to fix
<img width="379" alt="Screen Shot 2022-09-12 at 15 13 34" src="https://user-images.githubusercontent.com/7733167/189650529-05b7d5ea-11b9-4e60-b659-12ed143dafc3.png">
<img width="944" alt="Screen Shot 2022-09-12 at 15 13 15" src="https://user-images.githubusercontent.com/7733167/189650521-ebaa91e2-0990-4ea1-b47c-fd1ab68d8aff.png">
I'm saying this because we are trying to pixel-perfect based on the designs to eliminate 

As far as I see:
- hovering on the non-selected tab is missing
- there is a shadow difference 
- heights are different than designs
- right side of the edge of a Tab is transparent. Do we want that? We can ask Matt's opinion about this if needed.

ozgurkececioglu commented on packages/picasso/src/Tab/Tab.tsx:
```diff
@@ -73,6 +76,7 @@ export const Tab = forwardRef<HTMLDivElement, Props>(function Tab(props, ref) {
       onChange={onChange}
       onClick={onClick}
       classes={classes}
+      className={`PicassoTab-${orientation}`}
```
I wouldn't follow this approach. I think it would be better to use classes for the classname from `styles` file. Please check this example out for more info https://github.com//toptal/picasso/blob/2d76bbf2e1b565829145f3fd338ff86c9d198f6b/packages/picasso/src/Button/Button.tsx#L156

ozgurkececioglu commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -18,8 +18,17 @@ PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
+        borderRadius: '0.5rem 0 0 0.5rem',
+        margin: '0.125rem 0',
+        overflow: 'hidden',
+        padding: `${rem('8px')} 0 ${rem('8px')}`,
```
Is there a specific reason to not use `0.5rem` directly?

iatanas0v replied:
No, not really. I just tried to be a smartass, I guess :sweat_smile:

ozgurkececioglu commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -39,6 +48,20 @@ PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
       color: palette.common.black,
+
+      '&.PicassoTab-vertical': {
+        boxShadow: `0.25rem 0 0 ${palette.grey.lightest}, 0 0 0.25rem rgba(0, 0, 0, 0.08)`,
```
We have already defined boxShadows inside theme. Please check them out first before defining a custom one. And please make sure it is the same with figma 🙇‍♂️

iatanas0v replied:
I checked `packages/picasso-provider/src/Picasso/config/shadows.ts` but didn't find similar shadow. Is that the right place to look ?

Also you mentioned the shadow is different than the design. I double checked it and it's the very same :thinking: Maybe I'm missing something here, can you point out my mistake, please :pray:

ozgurkececioglu replied:
Hmm, I think you are right. We don't have this shadow in the codebase. But still, please define a variable and use it for `boxShadow` to increase readability 🙇‍♂️ 

About where to store this value (theme or just this file), we can discuss when others are reviewing this PR

iatanas0v commented:
> hovering on the non-selected tab is missing
fixed
> there is a shadow difference
see my other comment  https://github.com/toptal/picasso/pull/3096#discussion_r968475915
> heights are different than designs
fixed
> right side of the edge of a Tab is transparent. Do we want that? We can ask Matt's opinion about this if needed.
I didn't understand you :see_no_evil: :thinking: Can you elaborate, please

ozgurkececioglu commented on packages/picasso/src/Tab/Tab.tsx:
```diff
@@ -72,7 +76,14 @@ export const Tab = forwardRef<HTMLDivElement, Props>(function Tab(props, ref) {
       onClick={onClick}
-      classes={classes}
+      classes={{
+        root: cx({
+          [classes.horizontal]: orientation === 'horizontal',
+          [classes.vertical]: orientation === 'vertical',
+        }),
```
```suggestion
        root: classes[orientation],
```

ozgurkececioglu commented:
@iatanas0v 
I was referring to this part
<img width="379" alt="Screen Shot 2022-09-13 at 10 46 22" src="https://user-images.githubusercontent.com/7733167/189842140-61ea78a4-481a-4723-a447-db862725cc6d.png">

In the example stories, there is no edge. So, I'm having doubt that it should be stay like that or we should provide a default border 🤔

ozgurkececioglu commented on packages/picasso/src/Tabs/Tabs.tsx:
```diff
@@ -18,38 +18,51 @@ export interface Props
-    >
-      {children}
-    </MUITabs>
+    <TabsOrientationContext.Provider value={orientation || 'horizontal'}>
```
Again, I don't think covering the default value is not necessary. Wdyt?

iatanas0v replied:
In this case, it's here to please the TS. Another option is to use a not-null assertion :thinking: `value={orientation!}` as this should never be `null`. WDYT?

![image](https://user-images.githubusercontent.com/4757057/189844246-d678a2e7-f12d-4a26-8370-e8fc4cbac10b.png)

ozgurkececioglu replied:
Yeah, we generally use not-null assertion (until we decide to handle `defaultProps` in a better TS way). I think you can follow that for now

ozgurkececioglu commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -46,10 +42,53 @@ PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
+      borderRadius: `${sizes.borderRadius.medium} 0 0 ${sizes.borderRadius.medium}`,
+      margin: '0.5rem 0',
+      overflow: 'hidden',
+      padding: `${rem('9px')} 0 ${rem('9px')}`,
```
I think you made this change to make it 38px height exactly, right?

iatanas0v replied:
Yes, exactly. In the design there's less padding but the `line-height` of the label is `22px` and I saw `22px` is not something supported by `<Typography>`. I thought about using a custom css to force it to 22px but that would create visual inconsistency in cases of long labels (wrapping on few lines)

ozgurkececioglu commented on packages/picasso/src/Tabs/story/index.jsx:
```diff
@@ -26,6 +26,10 @@ page
   })
+  .addExample('Tabs/story/Vertical.example.tsx', {
+    title: 'Vertical',
+    takeScreenshot: false,
```
I think we can have this as a screenshot. We don't have vertical tabs in happo

ozgurkececioglu commented on packages/picasso/src/Tabs/test.tsx:
```diff
@@ -61,6 +62,21 @@ describe('Tabs', () => {
     expect(container).toMatchSnapshot()
   })
 
+  it('renders in vertical orientation', () => {
```
What is the purpose of this test? I mean, are you trying to test that none of the tabs are visible when there is no value?

iatanas0v replied:
I just wanted to have a snapshot test for the vertical option, that's all. I literally copied the above test case and changed orientation.

Looking now at it, I guess, we can strip it down to just `expect(container).toMatchSnapshot()`

ozgurkececioglu reviewed (approved):
Tested on my local. Looking good 👍 
Thanks for the contribution once again ❤️

denieler commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -1,8 +1,10 @@
+import { PicassoProvider, palette } from '@toptal/picasso-provider'
 
-PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
+const ACTIVE_SHADOW = `4px 0 0 ${palette.grey.lightest}, 0 0 4px rgba(0, 0, 0, 0.08)`
```
this doesn't seem right if this shadow is not taken from the `theme`

iatanas0v replied:
There's no similar shadow in the theme. Do you think we should add this one to the theme ? :thinking:

denieler replied:
@iatanas0v if this shadow is still not in Picasso - we need to check whether this shadow is in BASE. If yes - we should add this shadow to the list of other shadows in theme 👍 If not - we need to change this shadow to one, which is in the BASE most likely

iatanas0v replied:
@denieler After a chat with BASE it was decided to reuse an existing shadow: https://toptal-core.slack.com/archives/CCC3GP6CC/p1663224856285379?thread_ts=1662703900.679159&cid=CCC3GP6CC

denieler commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -46,10 +44,53 @@ PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
+      borderRadius: `${sizes.borderRadius.medium} 0 0 ${sizes.borderRadius.medium}`,
+      margin: '0.5rem 0',
+      overflow: 'hidden',
+      padding: `${rem('9px')} 0 ${rem('9px')}`,
```
can we use `em` here?

iatanas0v replied:
Sure. Just for the `padding`? Can you give me a quick explanation when do you use `rem` and when `em` so I know for the future :pray:

denieler replied:
Yes, we use `rem` only for `font-size` and `px` for `1px` cases (1px border, for example). All the rest places should use `em`

denieler commented on packages/picasso/src/Tab/styles.ts:
```diff
@@ -46,10 +44,53 @@ PicassoProvider.override(({ breakpoints, palette }: Theme) => ({
+    },
+    wrapper: {
+      '$vertical &': {
+        marginLeft: '1rem',
+        marginRight: '2rem',
```
can we move those styles to the `vertical` class?

OndrejTuma reviewed (approved):
I agree with Daniel about the standardized shadow,
otherwise it looks good to me 👍 
Thanks for contribution 🙇

denieler commented on packages/picasso/src/Tabs/styles.ts:
```diff
@@ -18,6 +18,15 @@ PicassoProvider.override(({ palette }: Theme) => ({
       },
     },
+    vertical: {
+      '&::after': {
+        display: 'none',
+      },
+
+      '& $indicator': {
+        display: 'none',
+      },
```
nit: idk, I have a bit mixed feelings about those classes, because it seems to me that the abstraction of `Tabs` is not good enough if in some special cases we need to hide some elements 🤔 usually it's better to do opposite - to show something based on a condition, instead of hiding 🤔

iatanas0v replied:
@denieler I hear you. I managed to avoid removing the "separator" line https://github.com/toptal/picasso/pull/3096/commits/6410c5ebb3c32c9004355565cd4ebeb564ed5e49 but I don't think it's easy to achieve the same for the indicator :unamused: 

I tried to reuse the indicator, just to move it to the left side (beginning of the tabs), but this way we can't match the design exactly as we can't have the exact same border radius and clip-effect from the tab :\

denieler reviewed (approved):
Haven't tested, but code looks good to me now 👍

iatanas0v commented:
@toptal/frontend-experience-eng can someone review and approve the Happo snapshot as well :pray:
